preflight_check/check.py

```diff
@@ -0,0 +1,246 @@
+        else:
+           task_inst.warn("Failed to detect firewall")
+
+
+def check_cluster():
+    """
+    A set of functions to check cluster state
+    """
+    print("\n============ Checking cluster state ============")
+    if not check_cluster_service():
+        return
+    check_fencing()
+    check_nodes()
+    check_resources()
+
+
+def check_cluster_service(quiet=False):
+    """
+    Check service status of pacemaker/corosync
+    """
+    task_inst = task.TaskCheck("Checking cluster service", quiet=quiet)
+    with task_inst.run():
+        if crmshutils.service_is_enabled("pacemaker"):
+            task_inst.info("pacemaker.service is enabled")
+        else:
+            task_inst.warn("pacemaker.service is disabled")
+
+        if crmshutils.service_is_enabled("corosync"):
+            task_inst.warn("corosync.service is enabled")
+
+        for s in ("corosync", "pacemaker"):
+            if crmshutils.service_is_active(s):
+                task_inst.info("{}.service is running".format(s))
+            else:
+                task_inst.error("{}.service is not running!".format(s))
+        return task_inst.passed
+
+
+def check_fencing():
+    """
+    Check STONITH/Fence:
+      Whether stonith is enabled
+      Whether stonith resource is configured and running
+    """
+    task_inst = task.TaskCheck("Checking STONITH/Fence")
+    with task_inst.run():
+        if not utils.FenceInfo().fence_enabled:
+            task_inst.warn("stonith is disabled")
+            return
+
+        task_inst.info("stonith is enabled")
+        rc, outp, _ = crmshutils.get_stdout_stderr("crm_mon -r1 | grep '(stonith:.*):'")
+        if rc != 0:
+            task_inst.warn("No stonith resource configured!")
+            return
+
+        res = re.search(r'([^\s]+)\s+\(stonith:(.*)\):\s+(\w+)', outp)
+        res_name, res_agent, res_state = res.groups()
+        common_msg = "stonith resource {}({})".format(res_name, res_agent)
+        state_msg = "{} is {}".format(common_msg, res_state)
+
+        task_inst.info("{} is configured".format(common_msg))
+        if res_state == "Started":
+            task_inst.info(state_msg)
+        else:
+            task_inst.warn(state_msg)
+
+        if re.search(r'sbd$', res_agent):
+            if crmshutils.service_is_active("sbd"):
+                task_inst.info("sbd service is running")
+            else:
+                task_inst.warn("sbd service is not running!")
+
+
+def check_nodes():
+    """
+    Check nodes info:
+      Current DC
+      Quorum status
+      Online/OFFLINE/UNCLEAN nodes
+    """
+    task_inst = task.TaskCheck("Checking nodes")
+    with task_inst.run():
+        rc, outp, errp = crmshutils.get_stdout_stderr("crm_mon -1")
+        if rc != 0:
+            task_inst.error("run \"crm_mon -1\" error: {}".format(errp))
+        # check DC
+        res = re.search(r'Current DC: (.*) \(', outp)
+        if res:
+            task_inst.info("DC node: {}".format(res.group(1)))
```

Comment:
How about use a function to parse all information(DC node, Online nodes, Offline nodes, Unclean nodes, etc...) of cluster, call the function to get information in everywhere.